File: src/main/java/com/dlut/algorithm/scala/Kmeans.scala
package com.dlut.algorithm.scala

import scala.collection.mutable
import scala.io.Source
import scala.util.Random

object Kmeans {
  val k = 2 //类个数
  val dim = 3 //数据集维度
  val shold = 0.0000000001 //阈值，用于判断聚类中心偏移量
  val centers = new Array[Vector[Double]](k) //聚类中心点（迭代更新）

  def main(args: Array[String]): Unit = {
    //处理输入数据
    val fileName = "D:\\kmeans_data.txt"
    val lines = Source.fromFile(fileName).getLines()
    val points = lines.map(line => { //数据预处理
      val parts = line.split(" ").map(_.toDouble)
      var vector = Vector[Double]()
      for (i <- 0 until dim)
        vector ++= Vector(parts(i))
      vector
    }).toArray
    initialCenters(points)
    kmeans(points, centers)
    printResult(points, centers)

  }

  //-------------------------------------随机初始化聚类中心---------------------------------------------------
  def initialCenters(points: Array[Vector[Double]]): Unit = {
    val pointsNum = points.length //数据集个数
    //寻找k个随机数(作为数据集的下标)
    val random = new Random()
    var index = 0
    var flag = true
    var temp = 0
    var array = new mutable.LinkedList[Int]() //保存随机下标号
    while (index < k) {
      temp = new Random().nextInt(pointsNum)
      flag = true
      if (array.contains(temp)) { //在数组中存在
        flag = false
      }
      else {
        if (flag) {
          array = array :+ temp
          index += 1
        }
      } //else-end
    } //while-end
    for (i <- 0 to centers.length - 1) {
      centers(i) = points(array(i))
      println("初始化中心点如下：")
      println(array(i))
      println(centers(i))
    }
  }

  //---------------------------迭代做聚类-------------------------------------
  def kmeans(points: Array[Vector[Double]], centers: Array[Vector[Double]]) = {
    var bool = true
    var newCenters = Array[Vector[Double]]()
    var move = 0.0
    var currentCost = 0.0 //当前的代价函数值
    var newCost = 0.0
    //根据每个样本点最近的聚类中心进行groupBy分组，最后得到的cluster是Map[Vector[Double],Array[Vector[Double]]]
    //Map中的key就是聚类中心，value就是依赖于该聚类中心的点集
    while (bool) { //迭代更新聚类中心，直到最优
      move = 0.0
      currentCost = computeCost(points, centers)
      val cluster = points.groupBy(v => closestCenter(centers, v))
      newCenters =
        centers.map(oldCenter => {
          cluster.get(oldCenter) match { //找到该聚类中心所拥有的点集
            case Some(pointsInThisCluster) =>
              //均值作为新的聚类中心
              vectorDivide(pointsInThisCluster.reduceLeft((v1, v2) => vectorAdd(v1, v2)), pointsInThisCluster.length)
            case None => oldCenter
          }
        })
      for (i <- 0 to centers.length - 1) {
        //move += math.sqrt(vectorDis(newCenters(i),centers(i)))
        centers(i) = newCenters(i)
      }
      /* if(move <= shold){
         bool = false
       }*/
      newCost = computeCost(points, centers) //新的代价函数值
      println("当前代价函数值：" + currentCost)
      println("新的代价函数值：" + newCost)
      if (math.sqrt(vectorDis(Vector(currentCost), Vector(newCost))) < shold)
        bool = false
    } //while-end
    println("寻找到的最优中心点如下：")
    for (i <- 0 to centers.length - 1) {
      println(centers(i))
    }
  }

  //--------------------------输出聚类结果-----------------------------
  def printResult(points: Array[Vector[Double]], centers: Array[Vector[Double]]) = {
    //将每个点的聚类中心用centers中的下标表示，属于同一类的点拥有相同的下标
    val pointsNum = points.length
    val pointsLabel = new Array[Int](pointsNum)
    var closetCenter = Vector[Double]()
    println("聚类结果如下：")
    for (i <- 0 to pointsNum - 1) {
      closetCenter = centers.reduceLeft((c1, c2) => if (vectorDis(c1, points(i)) < vectorDis(c2, points(i))) c1 else c2)
      pointsLabel(i) = centers.indexOf(closetCenter)
      println(points(i) + "-----------" + pointsLabel(i))
    }

  }

  //--------------------------找到某样本点所属的聚类中心-----------------------------
  def closestCenter(centers: Array[Vector[Double]], v: Vector[Double]): Vector[Double] = {
    centers.reduceLeft((c1, c2) =>
      if (vectorDis(c1, v) < vectorDis(c2, v)) c1 else c2
    )
  }

  //--------------------------计算代价函数（每个样本点到聚类中心的距离之和不再有很大变化）-----------------------------
  def computeCost(points: Array[Vector[Double]], centers: Array[Vector[Double]]): Double = {
    //cluster:Map[Vector[Double],Array[Vector[Double]]
    val cluster = points.groupBy(v => closestCenter(centers, v))
    var costSum = 0.0
    //var subSets = Array[Vector[Double]]()
    for (i <- 0 to centers.length - 1) {
      cluster.get(centers(i)) match {
        case Some(subSets) =>
          for (j <- 0 to subSets.length - 1) {
            costSum += (vectorDis(centers(i), subSets(j)) * vectorDis(centers(i), subSets(j)))
          }
        case None => costSum = costSum
      }
    }
    costSum
  }

  //--------------------------自定义向量间的运算-----------------------------
  //--------------------------向量间的欧式距离-----------------------------
  def vectorDis(v1: Vector[Double], v2: Vector[Double]): Double = {
    var distance = 0.0
    for (i <- 0 to v1.length - 1) {
      distance += (v1(i) - v2(i)) * (v1(i) - v2(i))
    }
    distance = math.sqrt(distance)
    distance
  }

  //--------------------------向量加法-----------------------------
  def vectorAdd(v1: Vector[Double], v2: Vector[Double]): Vector[Double] = {
    var v3 = v1
    for (i <- 0 to v1.length - 1) {
      v3 = v3.updated(i, v1(i) + v2(i))
    }
    v3
  }

  //--------------------------向量除法-----------------------------
  def vectorDivide(v: Vector[Double], num: Int): Vector[Double] = {
    var r = v
    for (i <- 0 to v.length - 1) {
      r = r.updated(i, r(i) / num)
    }
    r
  }
}

File: src/main/java/com/dlut/algorithm/KMeans.java
package com.dlut.algorithm;

import java.io.*;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.Map;
import java.util.Random;
import java.util.Vector;

public class KMeans {

    public int k;
    public int dim;
    public double shold;
    public ArrayList<double[]> centers;

    public KMeans(int k, int dim, double shold) {
        this.k = k;
        this.dim = dim;
        this.shold = shold;
        this.centers = new ArrayList<>();
    }

    public static void main(String[] args) throws IOException {
        /*double[] v1 = {1.1, 1.1, 1.1};
        double[] v2 = {1.2, 1.2, 1.2};
        double[] v3 = {1.3, 1.3, 1.3};
        double[] v4 = {4.1, 4.1, 4.1};
        double[] v5 = {4.2, 4.2, 4.2};
        double[] v6 = {4.3, 4.3, 4.3};

        ArrayList<double[]> points = new ArrayList<>();
        points.add(v1);
        points.add(v2);
        points.add(v3);
        points.add(v4);
        points.add(v5);
        points.add(v6);*/
        ArrayList<double[]> points = new ArrayList<>();

        KMeans kMeans = new KMeans(3, 3, 0.0000000001);
        points = kMeans.runKMeans("D:\\kmeans_data.txt");
        kMeans.initialCenters2(points);
        kMeans.kmeans(points);
        kMeans.printResult(points);

        // test closest
        /*double[] res = kMeans.closestCenter(v1);
        System.out.println("closest centser");
        System.out.println(res[0]);*/



    }

    public ArrayList<double[]> runKMeans(String path) throws IOException{
        File file = new File(path); //""

        if (file.isFile() && file.exists()){
            InputStreamReader reader = new InputStreamReader(Files.newInputStream(file.toPath()));
            BufferedReader bufferedReader = new BufferedReader(reader);
            String lineText = null;

            ArrayList<double[]> points = new ArrayList<>();

            while ((lineText = bufferedReader.readLine())!=null){
                double[] point = new double[dim];
//                System.out.println(lineText);
                String[] elems = lineText.split(" ");
                for (int i = 0; i<dim;i++){
                    String item = elems[i];
                    point[i] = Double.parseDouble(item);
                }
                points.add(point);
            }

//            System.out.println(points.size());

            return points;
        }
        return null;
    }

    // 初始化聚类中心
    public void initialCenters(ArrayList<double[]> points){
        int pointsNum = points.size();
        // random n
        Random random = new Random();
        int randn = random.nextInt(pointsNum);
        int[] centerIndex = new int[k];
        int pace = pointsNum / k;

        // 获得center下标
        for (int i = 0; i < k; i++){
            centerIndex[i] = (randn + i * pace) % pointsNum;
        }
        // 初始化中心
        for (int i : centerIndex) {
            centers.add(points.get(i));
        }
        // test
        /*System.out.println("centers:");
        for (double[] v: centers){
            System.out.println(v[0]);
        }*/
    }

    public void initialCenters2(ArrayList<double[]> points){
        int pointsNum = points.size();
        // random n
        Random random = new Random();
        int randn = random.nextInt(pointsNum);
        ArrayList<Integer> centerIndex = new ArrayList<>();
        centerIndex.add(randn);

        // 获得center下标 找k个点
        for (int i = 0; i < k - 1; i++){
            double maxDis = 0.0;
            int maxIndex = -1;
            // 遍历所有点
            for (int j = 0; j<pointsNum;j++){
                double dis = 0.0;
                // 遍历已选择的点
                for (int index : centerIndex){
                    dis += vectorDis(points.get(index), points.get(j));
                }
                if (dis > maxDis){
                    maxDis = dis;
                    maxIndex = j;
                }
            }
            // 找到最大的点 命名中心
            centerIndex.add(maxIndex);
        }
        // 初始化中心
        for (int i : centerIndex) {
            centers.add(points.get(i));
        }

        System.out.println("centers:");
        for (double[] v: centers){
            System.out.println(v[0]);
        }
    }

    //---------------------------迭代做聚类-------------------------------------
    public void kmeans(ArrayList<double[]> points){
        double currentCost = 0.0;
        double newCost = 0.0;
        ArrayList<double[]> newCenters = new ArrayList<>();

        double[][] newCenter = new double[k][dim];
        int[] newCenterCount = new int[k];


        while (true){
            System.out.println("start");
            for (int i=0;i<k;i++){
                for (int j = 0; j < dim; j ++){
                    newCenter[i][j] = 0.0;
                }
            }
            for (int i = 0; i < k; i++) {
                newCenterCount[i] = 0;
            }

            // 计算当前损失
            currentCost = computeCost(points);
            // 遍历所有点 找到点的中心点 所有一类的点坐标加起来
            for (double[] point: points){
                int centerIndex = closestCenter(point);
                for (int i = 0; i<dim;i++){
                    newCenter[centerIndex][i] += point[i];
                }
                newCenterCount[centerIndex] += 1;
            }
            // 计算新中心点
            for (int i=0;i<k;i++){
                double[] currCenter = centers.get(i);
                for (int j = 0; j < dim; j ++){
                    newCenter[i][j] = newCenter[i][j] / ((double)newCenterCount[i]);
                    currCenter[j] = newCenter[i][j];
                }

                centers.set(i, currCenter);
            }

            newCost = computeCost(points);

            // 判断迭代结束
            if (currentCost - newCost < shold){
                System.out.println("end");
                for (double[] center : centers){
                    System.out.println(center[0]);
                }
                break;
            }
        }
    }

    //--------------------------输出聚类结果-----------------------------
    public void printResult(ArrayList<double[]> points){
        for (double[] point: points){
            int centerIndex = closestCenter(point);
            for (int i = 0; i<dim; i++){
                System.out.printf(" %.1f ", point[i]);
            }
            System.out.printf(" center:%d \n", centerIndex);
        }
    }

    //--------------------------返回聚类结果（中心）-----------------------------
    public ArrayList<double[]> centerResult(){
        /*StringBuffer str = new StringBuffer();

        int count = 0;
        for (double[] center : centers){
            count++;
            System.out.println(center[0]);
            str.append("center: ");
            str.append(count);
            str.append("\n");
            for (int i=0; i<dim;i++){
                str.append(center[i]);
            }
        }

        System.out.println(str);

        return str;*/
        return centers;
    }

    //--------------------------计算代价函数-----------------------------
    public double computeCost(ArrayList<double[]> points){
        if (centers == null){
            System.out.println("no centers");
            return -1.0;
        }
        double dis = 0.0;
        for (double[] point: points){
            int centerIndex = closestCenter(point);
            double[] center = centers.get(centerIndex);
            dis += vectorDis(center, point);
        }
        return dis;
    }

    //--------------------------找到某样本点所属的聚类中心-----------------------------
    public int closestCenter(double[] v){
        double[] closestCenter = centers.get(0);
        int index = 0;
        double minDis = vectorDis(closestCenter, v);

        for (int i = 0 ; i < centers.size(); i++){
            double[] center = centers.get(i);
            double currDis = vectorDis(center, v);
            if (currDis < minDis){
                minDis = currDis;
                closestCenter = center;
                index = i;
            }
        }

        return index;
    }

    //--------------------------自定义向量间的运算-----------------------------
    //--------------------------向量间的欧式距离-----------------------------
    public double vectorDis(double[] v1, double[] v2){
        double distance = 0.0d;
        for (int i = 0; i < dim; i++) {
            distance += (v1[i] - v2[i]) * (v1[i] - v2[i]);
        }
        distance = Math.sqrt(distance);
        return distance;
    }

    //--------------------------向量加法-----------------------------
    public double[] vectorAdd(double[] v1, double[] v2){
        for (int i = 0; i < dim; i++) {
            v1[i] += v2[i];
        }
        return v1;
    }

    //--------------------------向量除法-----------------------------
    public double[] vectorDivide(double[] v, int num){
        for (int i = 0; i < dim; i++) {
            v[i] = v[i] / num;
        }
        return v;
    }

    public int getK() {
        return k;
    }

    public void setK(int k) {
        this.k = k;
    }

    public int getDim() {
        return dim;
    }

    public void setDim(int dim) {
        this.dim = dim;
    }

    public double getShold() {
        return shold;
    }

    public void setShold(double shold) {
        this.shold = shold;
    }

    public ArrayList<double[]> getCenters() {
        return centers;
    }

    public void setCenters(ArrayList<double[]> centers) {
        this.centers = centers;
    }
}
